### PR TITLE
[move-compiler-v2] Add enum scoping test and enhance match arm rewriting

### LINKED ISSUE #13794: [Bug][move-compiler-v2] ExpRewriterFunctions handling of Match isn't quite right
# 🐛 Bug

The `ExpRewriterFunctions::rewrite_exp_descent` handling of `Match` needs to call `rewrite_exit_scope` after calling `internal_rewrite_exp`.  Compare the code for `Block` with that for `Match`:
```
            Block(id, pat, binding, body) => {
                ...
                let (body_changed, new_body) = self.internal_rewrite_exp(body);
                self.rewrite_exit_scope(new_id);
                let newer_pat = if let Some(rewritten_pat) = optional_pat {
                ...
```
with
```
            Match(id, disc, arms) => {
                ...
                    let (body_changed, new_body) = self.internal_rewrite_exp(&arm.body);
                    new_arms.push(MatchArm {
```

The `Match` handler also ignores the result value when it calls `self.rewrite_enter_block_scope()`.  (Compare with `optional_pat = ...` in the `Body` clause).  Apparently that functionality was never use, as I see no current implementation of `ExpRewriterFunctions::rewrite_enter_block_scope()` returning anything other than `None`.  So we should simplify `rewrite_enter_block_scope` to not return a value and also simplify the `Block` clause.

The following example fails to generate the correct code due to the bug:
```
    /// Check for scoping issues
    public fun check_scoping(self: &Inner): u64 {
        let x = 3;
        {
            let x = 4;
            {
                match (self) {
                    Inner1{x: _} => true,
                    _ => false
                };
            };
        };
        x
    }
```

With the current code, the Simplifier causes this to yield code returning 4 instead of 3:
```
[variant baseline]
public fun m::check_scoping($t0: &m::Inner): u64 {
     var $t1: u64
     var $t2: bool
     var $t3: bool
     var $t4: &m::Inner
     var $t5: u64
  0: $t3 := test_variant m::Inner::Inner1($t0)
  1: if ($t3) goto 2 else goto 5
  2: label L2
  3: $t2 := true
  4: goto 12
  5: label L1
  6: $t4 := infer($t0)
  7: $t2 := false
  8: goto 12
  9: label L3
 10: $t5 := 15621340914461310977
 11: abort($t5)
 12: label L0
 13: $t1 := 4
 14: return $t1
}
```

### PR DESCRIPTION
**Pull request by brmataptos**
## Description

Fixes #13794

Trait `ExpRewriterFunctions` in `exp_rewriter.rs` didn't handle bind
scopes correctly in the case of `Match` arms: it was missing a call to
`rewrite_exit_scope`.  It also ignored the result of
`rewrite_enter_block_scope`, which currently doesn't have any
use cases that matter, but could in the future.  While fixing these, I added a new
`rewrite_match_arm` stub method for future potential use rewriting
match arms.

We might also later want a `rewrite_match` method, which would be
useful for cross-arm optimizations such as merging arms with identical
bodies or narrowing conditions in cases of overlaps, but I did not
include it here.

## How Has This Been Tested?

Added a test case (`enum_scoping.move`) which showed the bug due to
the AST constant propagation not handling scopes correctly in the case
of a Match expression.

## Key Areas to Review

One tricky issue is that match arms don't each have their own
`NodeId`, so we reuse the `Match`'s `id` for each of the arm scopes.

It's possible that some passes might use that `id` for something and
be broken.  That's not currently the case, as `ast_simplifier`
doesn't assume that variable scopes have unique ids across the whole
program, just within a single stack (and it's popped and reset between
arms now).  Other passes may need to be careful, unless we add a
`NodeId` for each arm.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by wrwg — APPROVED** [review]

File: third_party/move/move-model/src/exp_rewriter.rs [hunk 1]
```diff
@@ -224,6 +223,18 @@ pub trait ExpRewriterFunctions {
     ) -> Option<Exp> {
         None
     }
+    // Note that `rewrite_match_arm` is called *after* `rewrite_exit_scope`.
+    // (So all parameters here have already been processed.)
+    fn rewrite_match_arm(
+        &mut self,
+        id: NodeId, // id of the paremt match expression
```
Comment (vineethk) [inline third_party/move/move-model/src/exp_rewriter.rs:230]:
```suggestion
        id: NodeId, // id of the parent match expression
```

File: third_party/move/move-model/src/exp_rewriter.rs [hunk 2]
```diff
@@ -457,24 +456,39 @@ pub trait ExpRewriterFunctions {
             Match(id, disc, arms) => {
                 let (id_changed, new_id) = self.internal_rewrite_id(*id);
                 let (disc_changed, new_disc) = self.internal_rewrite_exp(disc);
+
```
Comment (vineethk) [inline third_party/move/move-model/src/exp_rewriter.rs:459]: nit: unintended newline?

**Review by vineethk — APPROVED** [review]
Nice!

File: third_party/move/move-model/src/exp_rewriter.rs (hunk 1 above)
Comment (brmataptos) [reply third_party/move/move-model/src/exp_rewriter.rs:230]: OK.

File: third_party/move/move-model/src/exp_rewriter.rs (hunk 2 above)
Comment (brmataptos) [reply third_party/move/move-model/src/exp_rewriter.rs:459]: I think having an empty line before the section dealing with arms made the code clearer, so I'm leaving it.

**Pull request merged**